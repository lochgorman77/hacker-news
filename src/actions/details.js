import { HackerNewsApi } from 'api/config';
import { detailsActions } from 'constants/ActionTypes';

export function fetchDetailsSuccess(detail) {
  return {
    type: detailsActions.FETCH_DETAILS_SUCCESS,
    payload: detail,
  };
}
export function fetchDetailsError(error) {
  return {
    type: detailsActions.FETCH_DETAILS_ERROR,
    payload: error,
  };
}
export function fetchDetails(detailId) {
  return function (dispatch) {
    dispatch({ type: detailsActions.FETCH_DETAILS });

    return HackerNewsApi.get(`/item/${detailId}.json`)
      .then((result) => {
        dispatch(fetchDetailsSuccess(result.data));
      })
      .catch((error) => {
        dispatch(fetchDetailsError(error));
      });
  };
}
