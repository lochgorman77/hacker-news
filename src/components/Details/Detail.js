import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { DateTime } from 'luxon';
import { withStyles } from '@material-ui/styles';
import { FormattedMessage } from 'react-intl';
import Card from '@material-ui/core/Card';
import CardHeader from '@material-ui/core/CardHeader';
import CardActions from '@material-ui/core/CardActions';
import CardContent from '@material-ui/core/CardContent';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';
import Link from '@material-ui/core/Link';

const styles = (theme) => ({
  root: {
    overflow: 'visible',
    minWidth: 500,
  },
  comment: {
    marginLeft: 'auto',
  },
});

const units = ['year', 'month', 'week', 'day', 'hour', 'minute', 'second'];

export class Detail extends Component {
  timeAgo = (date) => {
    const dateTime = DateTime.fromSeconds(date);
    const diff = dateTime.diffNow().shiftTo(...units);
    const unit = units.find((unit) => diff.get(unit) !== 0) || 'second';

    const relativeFormatter = new Intl.RelativeTimeFormat('en', {
      numeric: 'auto',
    });
    return relativeFormatter.format(Math.trunc(diff.as(unit)), unit);
  };

  render() {
    const { detail, classes } = this.props;

    const time = this.timeAgo(detail.time);
    return (
      <Card className={classes.root} raised>
        <CardHeader title={detail.title} />
        <CardContent>
          <Typography variant="caption" color="textSecondary">
            {detail.score} <FormattedMessage id={'detail.view.by'} /> {detail.by}{' '}
            {time.toLocaleString()}
          </Typography>
        </CardContent>
        <CardActions disableSpacing>
          <Button size="small">
            <Link href={detail.url} color="secondary" target="_blank" rel="noopener">
              <FormattedMessage id={'detail.view.link'} />
            </Link>
          </Button>
          <Button size="small" className={classes.comment}>
            {detail.descendants} <FormattedMessage id={'detail.view.comments'} />
          </Button>
        </CardActions>
      </Card>
    );
  }
}
Detail.propTypes = {
  detail: PropTypes.object,
  classes: PropTypes.object,
};

Detail.defaultProps = {
  detail: {},
};
export const DetailStyle = withStyles(styles, { name: 'Details' })(Detail);
export default DetailStyle;
