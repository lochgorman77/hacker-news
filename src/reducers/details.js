import { detailsActions } from './../constants/ActionTypes';

const initialState = {
  detail: [],
};

export default function roles(state = initialState, action) {
  switch (action.type) {
    case detailsActions.FETCH_DETAILS_SUCCESS:
      return {
        ...state,
        detail: [...state.detail, action.payload],
      };
    case detailsActions.FETCH_DETAILS_ERROR:
      return {
        ...state,
        detail: initialState.detail,
      };
    case detailsActions.RESET:
      return {
        ...state,
        detail: [],
      };
    default:
      return state;
  }
}
export const getDetails = (state) => state.details.detail;
