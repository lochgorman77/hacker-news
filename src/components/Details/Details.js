import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/styles';
import Grid from '@material-ui/core/Grid';
import Detail from 'components/Details/Detail';

const styles = (theme) => ({
  root: {
    overflow: 'visible',
  },
});

export class Details extends Component {
  render() {
    const { details, classes } = this.props;

    return (
      <Grid item container spacing={3}>
        {details.map((detail, index) => {
          if (detail !== null) {
            return (
              <Grid key={index} className={classes.card} item xs={6}>
                <Detail detail={detail} />
              </Grid>
            );
          } else {
            return null;
          }
        })}
      </Grid>
    );
  }
}
Details.propTypes = {
  details: PropTypes.array,
  classes: PropTypes.object,
};

Details.defaultProps = {
  details: {},
};
export const DetailsStyle = withStyles(styles, { name: 'Details' })(Details);
export default DetailsStyle;
