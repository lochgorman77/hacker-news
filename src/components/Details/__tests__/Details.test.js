import React from 'react';
import { DetailsStyle } from '../Details';
import details from 'utils/test-data/details-new.json';

const props = { details: [details] };

describe('<Detail />', () => {
  let shallow, wrapper;

  beforeEach(() => {
    shallow = createShallow({ dive: true });
    wrapper = shallow(<DetailsStyle {...props} />);
  });

  it('renders correctly', () => {
    expect(wrapper.length).toBe(1);
    expect(wrapper.name()).toBe('MuiGrid');
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
