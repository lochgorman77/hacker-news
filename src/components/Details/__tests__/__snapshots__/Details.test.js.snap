// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Detail /> renders correctly 1`] = `
<MuiGrid
  container={true}
  item={true}
  spacing={3}
>
  <MuiGrid
    item={true}
    key="0"
    xs={6}
  >
    <Details
      detail={
        Object {
          "by": "nojito",
          "descendants": 35,
          "id": 25428769,
          "kids": Array [
            25429182,
            25429065,
            25429334,
            25429105,
            25429145,
            25429074,
            25429035,
            25429242,
            25429234,
            25429067,
          ],
          "score": 143,
          "time": 1608027030,
          "title": "Appleâ€™s Anti-Tracking Plans for iPhone",
          "type": "story",
          "url": "https://foundation.mozilla.org/en/campaigns/apples-anti-tracking-plans-iphone/",
        }
      }
    />
  </MuiGrid>
</MuiGrid>
`;
